### PR TITLE
Syntax only server creates inferred project with all the open files w…

### PR DESCRIPTION
**Pull request by sheetalkamat**
…ith noResolve and can handle semantic operations

Fixes #

### CODE REVIEW

**Comment by sheetalkamat** [comment]
@typescript-bot pack this

**Comment by sheetalkamat** [comment]
@typescript-bot pack this

**Comment by sheetalkamat** [comment]
@mjbvz @minestarks @uniqueiniquity @amcasey @DanielRosenwasser @RyanCavanaugh 
https://github.com/microsoft/TypeScript/pull/38561#issuecomment-628308934 is build where syntaxOnly server can accept semantic commands.. ( i havent added any custom changes for any semantic commands nor have i limited any semantic operations as part of this prototype)

#38564 is PR that is on top of this and makes normal tsserver behave like syntaxOnly (one with semantic operations allowed)

**Comment by amcasey** [comment]
I'll be interested to see what this does to startup perf on the syntax server.  Looks pretty cool though.

**Comment by amcasey** [comment]
I've run into a pretty big issue on the VS side: semantic operations tend to be triggered for Roslyn Documents and those don't exist until the project is loaded.  So I see operations during project load _in already open documents_, but not in the newly opened document (i.e. the one that triggered the project load).  Obviously, this has nothing to do with the tsserver change.

**Comment by DanielRosenwasser** [comment]
@typescript-bot pack this

**Comment by DanielRosenwasser** [comment]
I'm not able to see things working better with

* the latest VS Code Insiders
* `"typescript.tsserver.useSeparateSyntaxServer": "dynamic",`
* using this build: https://github.com/microsoft/TypeScript/pull/38561#issuecomment-641468803

**Comment by DanielRosenwasser** [comment]
@typescript-bot pack this

**Comment by DanielRosenwasser** [comment]
I'm finally trying this out with VS Code Insiders! Pretty cool! A couple of thoughts and ideas

From our side:

- I think cross-file go-to-definition was one of the key scenarios we'd like to get to users, so I think we should really try to get that to light up, either here or in a follow-up PR.
- the fact that we show `any` when things aren't resolved in this mode is kind of confusing. For type display purposes, we might want to consider displaying the error type as something like a `loading` type.

@CyrusNajmabadi, how does Roslyn work when types haven't been resolved yet in things like quick info and signature help?

For example we have this signature

image

but `CodeActionsState.State` can't be resolved, so signature help looks like this:

Untitled
____

From the editor side:

- the startup time still isn't instantaneous for the syntax server, so I'm really never sure whether I can start editing reliably until I try quick info over and over. I think this is something we'll really need to provide a visual indication for before we make it broadly available.

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi, how does Roslyn work when types haven't been resolved yet in things like quick info and signature help?

We don't have that concept :)

File: src/server/session.ts [hunk 1]
```diff
@@ -575,6 +575,42 @@ namespace ts.server {
...
+        CommandNames.OpenExternalProject,
+        CommandNames.OpenExternalProjects,
+        CommandNames.CloseExternalProject,
+        CommandNames.SynchronizeProjectList,
+        CommandNames.EmitOutput,
+        CommandNames.CompileOnSaveAffectedFileList,
+        CommandNames.CompileOnSaveEmitFile,
+        CommandNames.CompilerOptionsDiagnosticsFull,
+        CommandNames.EncodedSemanticClassificationsFull,
+        CommandNames.SemanticDiagnosticsSync,
```
Comment (sheetalkamat) [inline src/server/session.ts:588]: SyntacticDiagnosticsSync: This is somewhat project setting dependent ? Atleast at some point parsing errors use to be different based on target .. eg target determines what unicode is considered identifier start...  So i am not sure if this should be enabled.. But if we do enable we i was wondering if GetErr should only do syntax checks and skip semantic and suggetion diagnostics on syntax server

Some questionable which i have disabled for now
Reload 
ReloadProjects
PrepareCallHierarchy
ProvideCallHierarchyIncomingCalls
ProvideCallHierarchyOutgoingCalls

**Comment by sheetalkamat** [comment]
@typescript-bot pack this

File: Gulpfile.js [hunk 2]
```diff
@@ -470,7 +470,7 @@ task("runtests").flags = {
     "   --shardId": "1-based ID of this shard (default: 1)",
 };
 
-const runTestsParallel = () => runConsoleTests("built/local/run.js", "min", /*runInParallel*/ cmdLineOptions.workers > 1, /*watchMode*/ false);
+const runTestsParallel =  async function () {};// () => runConsoleTests("built/local/run.js", "min", /*runInParallel*/ cmdLineOptions.workers > 1, /*watchMode*/ false);
```
Comment (amcasey) [inline Gulpfile.js:473]: Revert?

File: src/server/session.ts [hunk 3]
```diff
@@ -667,6 +703,15 @@ namespace ts.server {
             this.projectService = new ProjectService(settings);
             this.projectService.setPerformanceEventHandler(this.performanceEventHandler.bind(this));
             this.gcTimer = new GcTimer(this.host, /*delay*/ 7000, this.logger);
+
+            // Make sure to setup handlers to throw error for not allowed commands on syntax server;
+            if (this.projectService.syntaxOnly) {
+                invalidSyntaxOnlyCommands.forEach(commandName =>
+                    this.handlers.set(commandName, request => {
+                        throw new Error(`Request: ${request.command} not allowed on syntaxServer:: Request::${JSON.stringify(request)}`);
```
Comment (amcasey) [inline src/server/session.ts:711]: Why do we need the request?  In the log, it doesn't matter if it's well-formed - it's that it got sent at all.  It'll be immediately before the error in the log anyway, right?

File: src/services/services.ts [hunk 4]
```diff
@@ -1171,6 +1171,26 @@ namespace ts {
         }
     }
 
+    const invalidOperationsOnSyntaxOnly: readonly (keyof LanguageService)[] = [
```
Comment (amcasey) [inline src/services/services.ts:1174]: How does this relate to the list in session.ts?  Do they need to stay in sync?  Is one redundant?

File: src/services/services.ts [hunk 5]
```diff
@@ -2259,6 +2272,16 @@ namespace ts {
             provideCallHierarchyIncomingCalls,
             provideCallHierarchyOutgoingCalls
         };
+
+        if (syntaxOnly) {
+            invalidOperationsOnSyntaxOnly.forEach(key =>
+                ls[key] = (...args: any[]) => {
+                    throw new Error(`LanguageService Operation: ${key} not allowed on syntaxServer:: arguments::${JSON.stringify(args)}`);
```
Comment (amcasey) [inline src/services/services.ts:2279]: As above, we probably don't need the arguments.

File: src/server/session.ts (hunk 3 above)
Comment (amcasey) [reply src/server/session.ts:711]: Do we want to test this rejection mechanism?  Not for all commands, but maybe for one?

File: Gulpfile.js (hunk 2 above)
Comment (sheetalkamat) [reply Gulpfile.js]: Thanks for catching this./. I thought i did revert but had missed it.

File: src/server/session.ts (hunk 3 above)
Comment (sheetalkamat) [reply src/server/session.ts:711]: Already there https://github.com/microsoft/TypeScript/pull/38561/files#diff-0403b837aaec2ba7b3953efef687f19aR66 tests both LS and session.
This is to ensure that plugins or anyone doesnt accidently use semantic operation on LS or session from API point of view

**Comment by DanielRosenwasser** [comment]
Speaking of logs- do editors log from the syntax server?

**Comment by amcasey** [comment]
VS logging is opt-in (via env var or regkey), but, yes, it applies to both.

VS Code only has a command to show the semantic log (AFAIK), but I'm pretty sure it enables both.

**Comment by sheetalkamat** [comment]
@mjbvz We definitely want users to be able to share the syntaxOnly server logs easily.

**Pull request merged**